Proceso:

● Transformación y limpieza de los datos

● Análisis descriptivo de los datos

● Visualización de los datos

● Informe explicativo del análisis


In [9]:
import pandas as pd

#para que aparezcan todas las columnas
pd.set_option('display.max_columns', None)

import numpy as np

In [10]:
datos_bancarios_original=pd.read_csv('bank-additional.csv', index_col = 0)
datos_bancarios_original.sample(5)

age          job   marital            education  default  housing  \
10533  36.0   technician   MARRIED  professional.course      0.0      1.0   
849     NaN   management  DIVORCED    university.degree      0.0      1.0   
21426   NaN   management   MARRIED          high.school      0.0      1.0   
28457  45.0  blue-collar   MARRIED             basic.9y      NaN      1.0   
33466  22.0  blue-collar   MARRIED             basic.9y      0.0      NaN   

       loan    contact  duration  campaign  pdays  previous     poutcome  \
10533   0.0  telephone       133         1    999         0  NONEXISTENT   
849     0.0  telephone       236         1    999         0  NONEXISTENT   
21426   0.0   cellular       173         1    999         0  NONEXISTENT   
28457   0.0   cellular       427         4    999         1      FAILURE   
33466   NaN   cellular       453         2    999         0  NONEXISTENT   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed   y  \
10533           1.4         94,465         -41,8     4,961      5228,1  no   
849             1.1         93,994         -36,4     4,856        5191  no   
21426           1.4         93,444         -36,1     4,963      5228,1  no   
28457          -1.8         93,075         -47,1     1,423      5099,1  no   
33466          -1.8         92,893         -46,2     1,291      5099,1  no   

                   date  latitude  longitude  \
10533    1-febrero-2015    44.575    -90.749   
849    8-diciembre-2017    28.251    -79.270   
21426     24-abril-2017    30.078    -75.854   
28457     28-marzo-2017    39.090    -88.552   
33466      7-julio-2016    35.457    -95.715   

                                        id_  
10533  3c33e0da-d2b8-4a6e-b139-73f6eaa31054  
849    2c5b47ee-0c81-48e4-ab89-4c85568cec39  
21426  07e1c80e-9d7a-4aed-9df5-e2d747b33f35  
28457  d1d02eb5-f24e-4f43-8f95-a7b9a3e9b393  
33466  edf85771-1e96-4f2d-a767-201366f02b0a

#### Limpieza datos bancarios:
- indice en columa en blanco igual a unnamed: 0 ---> index_col=0 al importar
- defaults tiene bastantes NaN
- housing y loan están con decimales: Sí:1, No:0.
- mayusculas y minusculas en varias columnas
- mezcla de puntos y comas en números
- fecha tiene todo conjunto separado por un guion
- age con NaN

### Columnas
- **age**: como float --> cambiar a integer. Tiene nulos (11.91%)
-**job**: al final tiene punto, cambiado a _. Quitarlo al estar mal escrito
- **marital** texto en mayus -->cambiar a lower. Cambiado!
- **education**: tiene nulos NaN. palabras separadas por puntos en lugar de espacios. cambiar "." por " " texto cambiado a lower sin espacios!
- **default**: como float64 ---> cambiar a booleano o texto al ser 1= sí, 0 =no. tiene nulos (20.89%)
- **housing**: como float64 ---> cambiar a booleano o texto al ser 1= sí, 0 =no. Tiene nulos (2.39%)
- **loan**: como float64 ---> cambiar a booleano o texto al ser 1= sí, 0 =no. Tiene nulos (2.39%)
- **duration** tiene un minimo de 0 segundos en la llamada, error? llamada no recibida por ese cliente?
- **poutcome**: en mayuscula --cambiar a lower. cambiado!
- **cons.price.idx**: object-->cambio a float. tiene coma, los float se consideran con los numeros con . decimal. cambiar coma por . Tiene nulos (1.10%) Cambiado a float
- **cons.conf.idx**: object-->cambio a float. tiene coma, los float se consideran con los numeros con . decimal. cambiar coma por . Cambiado a float
- **euribor**: como object--> cambiar a float. tiene nulos (21.53%). Cambiado a float
- **nr.employed**: como object, es un numero con una coma, pasar a float? cambiar coma por .
- **date**: como object --> pasar a fecha. Tiene nulos (0.58%). Pasado a fechas

### ¿Eliminacion columnas redundantes?
Para comprobar si algunas columnas son iguales:
````python
dataframe_importado['etiqueta_columna'].equals(dataframe_importado['etiqueta_columna_a_comparar'])
````
Devuelve booleano, si es que sí (True), la columna A es igual a la B. Si da No(false), no son iguales
En caso afirmativo, eliminamos en función de cual tiene más coherencia con el resto de las columnas. ej_que nombre de columnase parece más, texto minuscula, separación por guiones...

Para eliminar la columna:
````python
dataframe_importado.drop(etiqueta_columna_a_eliminar, axis=1)
````
axis=1 para tener en cuenta que es una columna.

Ojo: al hacer este calculo no se modifica el dataframe en el que estamos trabajando, para modificar este se renombra la variable:
````python
dataframe_importado=dataframe_importado.drop(etiqueta_columna_a_eliminar, axis=1)
````
o directamente el parametro:
````python
dataframe_importado.drop(etiqueta_columna_a_eliminar, axis=1, inplace= True)

No cambiar nunca el original datos_bancarios_original, dejarlo en raw
````

In [11]:
#Hacemos copia del documento original
datos_bancarios=datos_bancarios_original.copy()

In [12]:
#para ver las filas (registros) con sus columnas
datos_bancarios.shape

(43000, 23)

In [13]:
f'Tenemos un total de {datos_bancarios.shape[0]} filas y  {datos_bancarios.shape[1]} columnas'

'Tenemos un total de 43000 filas y  23 columnas'

In [14]:
#numero de entradas, indices, numero columnas, indices columnas, nombres columnas, nulos y tipos de las columnas
#ver que tipos coinciden y no con la tabla de muestra
datos_bancarios.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             37880 non-null  float64
 1   job             42655 non-null  object 
 2   marital         42915 non-null  object 
 3   education       41193 non-null  object 
 4   default         34019 non-null  float64
 5   housing         41974 non-null  float64
 6   loan            41974 non-null  float64
 7   contact         43000 non-null  object 
 8   duration        43000 non-null  int64  
 9   campaign        43000 non-null  int64  
 10  pdays           43000 non-null  int64  
 11  previous        43000 non-null  int64  
 12  poutcome        43000 non-null  object 
 13  emp.var.rate    43000 non-null  float64
 14  cons.price.idx  42529 non-null  object 
 15  cons.conf.idx   43000 non-null  object 
 16  euribor3m       33744 non-null  object 
 17  nr.employed     43000 non-null  obje

In [15]:
#para ver los nulos de forma más fiable que con info:

datos_bancarios.isnull().sum()

age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

In [16]:
#para saber el porcentaje de nulos en relación al total

round(datos_bancarios.isnull().mean()*100,2)

age               11.91
job                0.80
marital            0.20
education          4.20
default           20.89
housing            2.39
loan               2.39
contact            0.00
duration           0.00
campaign           0.00
pdays              0.00
previous           0.00
poutcome           0.00
emp.var.rate       0.00
cons.price.idx     1.10
cons.conf.idx      0.00
euribor3m         21.53
nr.employed        0.00
y                  0.00
date               0.58
latitude           0.00
longitude          0.00
id_                0.00
dtype: float64

In [17]:
#Comprobamos si hay duplicados
datos_bancarios.duplicated().sum()

#Da cero, no hay duplicados

np.int64(0)

In [18]:
#para ver las estadisticas principales, en formato T para hcerlo mas sencillo de ver
datos_bancarios.describe().T

count        mean         std      min        25%       50%  \
age           37880.0   39.977112   10.437957   17.000   32.00000   38.0000   
default       34019.0    0.000088    0.009390    0.000    0.00000    0.0000   
housing       41974.0    0.535998    0.498708    0.000    0.00000    1.0000   
loan          41974.0    0.155620    0.362499    0.000    0.00000    0.0000   
duration      43000.0  257.739279  258.666033    0.000  102.00000  179.0000   
campaign      43000.0    2.567233    2.772294    1.000    1.00000    2.0000   
pdays         43000.0  962.330953  187.260394    0.000  999.00000  999.0000   
previous      43000.0    0.174023    0.497366    0.000    0.00000    0.0000   
emp.var.rate  43000.0    0.077128    1.573898   -3.400   -1.80000    1.1000   
latitude      43000.0   36.856697    7.225948   24.396   30.61475   36.7610   
longitude     43000.0  -95.939067   16.752282 -124.997 -110.49425  -95.8995   

                    75%       max  
age            47.00000    98.000  
default         0.00000     1.000  
housing         1.00000     1.000  
loan            0.00000     1.000  
duration      319.00000  4918.000  
campaign        3.00000    56.000  
pdays         999.00000   999.000  
previous        0.00000     7.000  
emp.var.rate    1.40000     1.400  
latitude       43.11325    49.384  
longitude     -81.42775   -66.937

###Primeras conclusiones de estadisticas principales:
- **media de edad** 40 años, (cuidado, hay  11.9% nulos), con el mas joven teniendo 17 años, maximo 98
- **loan**: Tienden a no tener otro tipo de prestamos. (media: 0.15, más cercana a 0 que a 1)
-**duration**: Media de interaccion con el cliente de 257 segundos (257/60: media de 4,2 minutos). Maxima duración de 1 hora 36 minutos, Mínimo de 0 segundos??
-**campaign**:  de media, se han realizado casi 3 contactos con cliente durante esta campaña (media 2.5)

In [19]:
257/60

4.283333333333333

In [20]:
#llamada maxima de 4918 segundos, 81 minutos, 1 hora y 36 minutos
print(4918/60)
print(81.96/60)

81.96666666666667
1.3659999999999999


In [21]:
#datos unicos para default, comprobamos si hay más numeros aparte de los válidos: 0 y 1
datos_bancarios['default'].unique()

array([ 0., nan,  1.])

In [22]:
#datos unicos para housing, comprobamos si hay más numeros aparte de los válidos: 0 y 1
datos_bancarios['housing'].unique()

array([ 0.,  1., nan])

In [24]:
#datos unicos para loan, comprobamos si hay más numeros aparte de los válidos: 0 y 1
datos_bancarios['loan'].unique()

array([ 0.,  1., nan])

In [25]:
#hacer value count de cada columna de categoria: marital, education
#hacer limpieza y cambiar, hay muchas columnas como object que no lo son realmente

for col in datos_bancarios.select_dtypes(include='O').columns:
    print(col.upper())
    print(datos_bancarios[col].value_counts())
    print('---------------')

JOB
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
---------------
MARITAL
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
---------------
EDUCATION
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
---------------
CONTACT
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
---------------
POUTCOME
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64
---------------
CONS.PRICE.IDX
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4

## Homogeneizamos texto en columnas categoricas

In [ ]:
#para poner todas las columnas con string en minuscula, reemplazar a barra baja los espacios y separaciones por puntos

for col in datos_bancarios.select_dtypes(include='O').columns:
    datos_bancarios[col]= datos_bancarios[col].str.lower()

In [33]:
#para poner todas las columnas con string con espacio en lugar de punto
for col in datos_bancarios.select_dtypes(include='object').columns:
    datos_bancarios[col]= datos_bancarios[col].str.replace(".", "_")

#equivalente en excel de hacer operacion con una columna, se elimina la anterior y se renombra con el nombre de la columna eliminada.

In [65]:
#comprobamos que esté en minúsculas y separado por espacios
datos_bancarios.sample(1)

age       job  marital    education  default  housing  loan    contact  \
5590  37.0  services  married  high_school      0.0      0.0   0.0  telephone   

      duration  campaign  pdays  previous     poutcome  emp.var.rate  \
5590        63         1    999         0  nonexistent           1.1   

     cons.price.idx cons.conf.idx euribor3m nr.employed   y       date  \
5590         93,994         -36,4     4,857        5191  no 2015-03-12   

      latitude  longitude                                   id_  
5590    41.212   -110.362  62adb5d1-d943-4883-8d6f-214e319d6813

## Pasar datos a fecha

In [34]:
#panda tiene instalado datetime, pueden utilizarse los metodos directamente

In [64]:
datos_bancarios.sample(2)

age         job   marital            education  default  housing  \
22538  34.0  technician  divorced  professional_course      0.0      0.0   
32659  32.0  technician    single  professional_course      0.0      1.0   

       loan    contact  duration  campaign  pdays  previous     poutcome  \
22538   0.0  telephone       336         1    999         0  nonexistent   
32659   0.0   cellular       183         1    999         0  nonexistent   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed   y  \
22538           1.4         93,444         -36,1       NaN      5228,1  no   
32659          -1.8         92,893         -46,2     1,299      5099,1  no   

            date  latitude  longitude                                   id_  
22538 2017-02-24    34.189    -90.580  654ed50e-e2f3-491a-8496-41325272e674  
32659 2019-02-21    34.358    -77.682  9e6f82bc-e496-4cf6-aa28-59b9ed4c0b8a

In [48]:
#al estar en español, necesitamos cambiar el mes a inglés. Uso de diccionarios:

meses_es = {
    "enero": "01",
    "febrero": "02",
    "marzo": "03",
    "abril": "04",
    "mayo": "05",
    "junio": "06",
    "julio": "07",
    "agosto": "08",
    "septiembre": "09",
    "setiembre": "09",  # variante válida en español
    "octubre": "10",
    "noviembre": "11",
    "diciembre": "12"
}


In [59]:
datos_bancarios.replace({'date':meses_es}, regex=True, inplace=True)
datos_bancarios.sample(5)

age     job   marital          education  default  housing  loan  \
31939  28.0  admin_    single        high_school      NaN      0.0   0.0   
33741  33.0  admin_  divorced  university_degree      0.0      1.0   0.0   
30745  35.0  admin_    single  university_degree      0.0      1.0   0.0   
20648   NaN  admin_  divorced  university_degree      0.0      0.0   0.0   
13096  39.0  admin_   married  university_degree      0.0      1.0   1.0   

         contact  duration  campaign  pdays  previous     poutcome  \
31939   cellular      1391         2    999         0  nonexistent   
33741  telephone        47         1    999         0  nonexistent   
30745  telephone       227         3    999         0  nonexistent   
20648   cellular       185         1    999         0  nonexistent   
13096   cellular       184         1    999         0  nonexistent   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed    y  \
31939          -1.8         92,893         -46,2     1,327      5099,1  yes   
33741          -1.8         92,893         -46,2       NaN      5099,1   no   
30745          -1.8         92,893         -46,2     1,344      5099,1   no   
20648           1.4         93,444         -36,1       NaN      5228,1   no   
13096           1.4         93,918         -42,7     4,962      5228,1   no   

             date  latitude  longitude                                   id_  
31939  19-11-2019    48.251   -121.610  082563e2-87c8-4845-9b9c-06aa2aa5733d  
33741  23-11-2018    33.583   -116.592  68b6ba4a-b562-4192-8c8b-8076c3903b06  
30745  15-07-2017    46.006   -113.017  6b17b221-9e55-4d47-a720-bf9d2c5960d4  
20648   3-08-2015    46.628   -120.909  6aa2bb82-c678-4725-94f8-1245fcedd8b6  
13096   7-02-2019    42.455    -90.081  72a59895-405a-4702-8109-418173604476

In [62]:
#una vez cambiado de español a imglés, lanzar para convertir a fecha
datos_bancarios['date'] = pd.to_datetime(datos_bancarios['date'], format='%d-%m-%Y')

In [63]:
datos_bancarios.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   age             37880 non-null  float64       
 1   job             42655 non-null  object        
 2   marital         42915 non-null  object        
 3   education       41193 non-null  object        
 4   default         34019 non-null  float64       
 5   housing         41974 non-null  float64       
 6   loan            41974 non-null  float64       
 7   contact         43000 non-null  object        
 8   duration        43000 non-null  int64         
 9   campaign        43000 non-null  int64         
 10  pdays           43000 non-null  int64         
 11  previous        43000 non-null  int64         
 12  poutcome        43000 non-null  object        
 13  emp.var.rate    43000 non-null  float64       
 14  cons.price.idx  42529 non-null  object        
 15  cons.co

## Cambiar comas por puntos

In [66]:
for col in datos_bancarios.select_dtypes(include='O').columns:
    datos_bancarios[col]= datos_bancarios[col].str.replace(',','.')
    try:
        datos_bancarios[col]=datos_bancarios[col].astype(float)
    except:
        pass

In [67]:
datos_bancarios.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   age             37880 non-null  float64       
 1   job             42655 non-null  object        
 2   marital         42915 non-null  object        
 3   education       41193 non-null  object        
 4   default         34019 non-null  float64       
 5   housing         41974 non-null  float64       
 6   loan            41974 non-null  float64       
 7   contact         43000 non-null  object        
 8   duration        43000 non-null  int64         
 9   campaign        43000 non-null  int64         
 10  pdays           43000 non-null  int64         
 11  previous        43000 non-null  int64         
 12  poutcome        43000 non-null  object        
 13  emp.var.rate    43000 non-null  float64       
 14  cons.price.idx  42529 non-null  float64       
 15  cons.co

## Cambio admin en job

In [79]:
datos_bancarios['job']=datos_bancarios['job'].str.replace("_", "")

In [82]:
datos_bancarios.sample(5)

age           job  marital          education  default  housing  loan  \
20961  31.0         admin   single  university_degree      0.0      1.0   0.0   
10550  24.0   blue-collar  married           basic_9y      NaN      0.0   0.0   
25872  54.0  entrepreneur  married  university_degree      0.0      1.0   0.0   
25781  31.0    technician  married           basic_9y      0.0      1.0   0.0   
13267  26.0         admin  married        high_school      0.0      1.0   1.0   

         contact  duration  campaign  pdays  previous     poutcome  \
20961   cellular       108         1    999         0  nonexistent   
10550  telephone       218         1    999         0  nonexistent   
25872   cellular       159         2      4         1      success   
25781   cellular        17         5    999         1      failure   
13267   cellular       273         5    999         0  nonexistent   

       emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed  \
20961           1.4          93.444          -36.1      4.964       5228.1   
10550           1.4          94.465          -41.8      4.961       5228.1   
25872          -0.1          93.200          -42.0      4.120       5195.8   
25781          -0.1          93.200          -42.0      4.120       5195.8   
13267           1.4          93.918          -42.7      4.962       5228.1   

        y       date  latitude  longitude  \
20961  no 2017-08-31    30.604    -73.265   
10550  no 2015-03-19    31.066   -123.981   
25872  no 2015-11-30    42.112    -71.099   
25781  no 2016-03-29    38.319   -116.605   
13267  no 2018-07-03    29.040    -96.926   

                                        id_  
20961  264cd29f-4d24-4ba3-94bb-5b0c5a07ff4e  
10550  445fc6e4-257e-4af8-8aec-0a99f3037d9e  
25872  93d6e382-ac93-40a1-9617-44766fbf2310  
25781  d05a678a-94d4-4849-b2a5-90eae777cab3  
13267  3048f5b9-43fb-4a44-8d40-4a4ef52e4e53

## Cambio a booleanos

In [ ]:
datos_cliente = pd.read_excel('customer-details.xlsx')
datos_cliente.sample(5)